### PR TITLE
Add block issuer key source for implicit account transition

### LINKED ISSUE #1666: Add more PublicKey generation fallbacks
- [x] https://github.com/iotaledger/iota-sdk/pull/1598/files#diff-bb97bc42ed78814b7c87abe4a31292fcfa12055d3bf69e2dd7ed05666178fbb0R51

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Adds block issuer key source override to implicit account transition which can be a key or a bip path.

## Links to any relevant issues

Closes #1666

### CODE REVIEW

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/core/src/method/wallet.rs [hunk 1]
```diff
@@ -194,7 +195,11 @@ pub enum WalletMethod {
     /// Prepares to transition an implicit account to an account.
     /// Expected response: [`PreparedTransaction`](crate::Response::PreparedTransaction)
     #[serde(rename_all = "camelCase")]
-    PrepareImplicitAccountTransition { output_id: OutputId },
+    PrepareImplicitAccountTransition {
+        output_id: OutputId,
+        public_key: Option<PublicKey>,
+        bip_path: Option<Bip44>,
```
Comment (Alex6323) [inline bindings/core/src/method/wallet.rs]: So if `public_key` is set the value in `bip_path` is ignored no matter if it's Some or None?

Could this not already be using the enum:
```rust
key_source: Option<BlockIssuerKeySource>
```

Comment (DaughterOfMars) [reply bindings/core/src/method/wallet.rs]: Yeah, that is correct. I didn't want to make that utility enum public really, and this seemed easier for users. I forgot to put default tags on them though.

Comment (Alex6323) [reply bindings/core/src/method/wallet.rs]: okay :+1:

**Review by Alex6323 — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/core/src/method_handler/wallet.rs [hunk 2]
```diff
@@ -210,8 +210,18 @@ pub(crate) async fn call_wallet_method_internal(wallet: &Wallet, method: WalletM
...
+            output_id,
+            public_key,
+            bip_path,
+        } => {
+            let data = if public_key.is_some() {
+                wallet
+                    .prepare_implicit_account_transition(&output_id, public_key)
+                    .await?
+            } else {
+                wallet.prepare_implicit_account_transition(&output_id, bip_path).await?
```
Comment (thibault-martinez) [inline bindings/core/src/method_handler/wallet.rs:220]: Is there nothing we can do with an Option method to do one or the other without the if ?

File: bindings/core/src/method/wallet.rs [hunk 3]
```diff
@@ -194,7 +195,13 @@ pub enum WalletMethod {
     /// Prepares to transition an implicit account to an account.
     /// Expected response: [`PreparedTransaction`](crate::Response::PreparedTransaction)
     #[serde(rename_all = "camelCase")]
-    PrepareImplicitAccountTransition { output_id: OutputId },
+    PrepareImplicitAccountTransition {
+        output_id: OutputId,
+        #[serde(default)]
+        public_key: Option<PublicKey>,
```
Comment (thibault-martinez) [inline bindings/core/src/method/wallet.rs]: Since this is not our type, how does this serialise ? prefixed hex ?

File: cli/src/wallet_cli/mod.rs [hunk 4]
```diff
@@ -584,8 +586,18 @@ pub async fn implicit_account_creation_address_command(wallet: &Wallet) -> Resul
...
+    wallet: &Wallet,
+    output_id: OutputId,
+    public_key: Option<String>,
+) -> Result<(), Error> {
+    let public_key = public_key
+        .map(|s| {
+            PublicKey::try_from_bytes(prefix_hex::decode(s).map_err(|e| Error::Miscellaneous(e.to_string()))?)
+                .map_err(|e| Error::Miscellaneous(e.to_string()))
+        })
+        .transpose()?;
```
Comment (thibault-martinez) [inline cli/src/wallet_cli/mod.rs:598]: No way to make this prettier 😭 ?

File: sdk/src/wallet/operations/transaction/account.rs [hunk 5]
```diff
@@ -81,3 +103,9 @@ where
             .await
     }
 }
+
+#[derive(From)]
+pub enum BlockIssuerKeySource {
+    Key(PublicKey),
+    Path(Bip44),
```
Comment (thibault-martinez) [inline sdk/src/wallet/operations/transaction/account.rs]: I'd maybe call the variant BipPath, Path is a bit too generic IMO

File: bindings/core/src/method_handler/wallet.rs (hunk 2 above)
Comment (DaughterOfMars) [reply bindings/core/src/method_handler/wallet.rs:220]: no, because the static typing is different

File: bindings/core/src/method/wallet.rs (hunk 3 above)
Comment (DaughterOfMars) [reply bindings/core/src/method/wallet.rs]: HMMMM I doubt it. probably as bytes. What do we want?

File: cli/src/wallet_cli/mod.rs (hunk 4 above)
Comment (DaughterOfMars) [reply cli/src/wallet_cli/mod.rs:598]: I can try I guess, but I don't think so

File: sdk/src/wallet/operations/transaction/account.rs (hunk 5 above)
Comment (DaughterOfMars) [reply sdk/src/wallet/operations/transaction/account.rs]: alrighty

File: bindings/core/src/method/wallet.rs (hunk 3 above)
Comment (thibault-martinez) [reply bindings/core/src/method/wallet.rs]: prefixed_hex I'd say, but maybe this is already the case, have to check

File: bindings/core/src/method_handler/wallet.rs (hunk 2 above)
Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs:220]: sad reacts only

File: bindings/core/src/method_handler/wallet.rs [hunk 6]
```diff
@@ -204,8 +205,20 @@ pub(crate) async fn call_wallet_method_internal(wallet: &Wallet, method: WalletM
...
-        WalletMethod::PrepareImplicitAccountTransition { output_id } => {
-            let data = wallet.prepare_implicit_account_transition(&output_id).await?;
+        WalletMethod::PrepareImplicitAccountTransition {
+            output_id,
+            public_key,
+            bip_path,
+        } => {
+            let data = if let Some(public_key_str) = public_key {
+                let public_key = PublicKey::try_from_bytes(prefix_hex::decode(public_key_str)?)
+                    .map_err(iota_sdk::wallet::Error::other)?;
```
Comment (thibault-martinez) [inline bindings/core/src/method_handler/wallet.rs]: Why is this needed? Wallet has a From impl for crypto errors.

Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs]: Ah wait PublicKey is our type

Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs]: Hmm I think I would add a Types variant in Wallet + a from tbh, seems like something useful

Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs]: Ahhh there is `Block` but no from for it, maybe just add it?

File: bindings/core/src/method/wallet.rs (hunk 3 above)
Comment (DaughterOfMars) [reply bindings/core/src/method/wallet.rs]: done

File: cli/src/wallet_cli/mod.rs (hunk 4 above)
Comment (DaughterOfMars) [reply cli/src/wallet_cli/mod.rs:598]: it cannot

Comment (thibault-martinez) [reply cli/src/wallet_cli/mod.rs:598]: Alright, thanks

File: bindings/core/src/method_handler/wallet.rs (hunk 6 above)
Comment (DaughterOfMars) [reply bindings/core/src/method_handler/wallet.rs]: I'm confused, you mean add an error variant?

Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs]: I'm confused as well, why can't we just `?` since the wallet error has a from impl for crypto error ?

Comment (DaughterOfMars) [reply bindings/core/src/method_handler/wallet.rs]: it didn't help much but I fixed it

Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs]: But for my understanding, why can't we use ? directly ?

Comment (DaughterOfMars) [reply bindings/core/src/method_handler/wallet.rs]: Because the error type returned here is local to the bindings, not the wallet error

**Review by thibault-martinez — COMMENTED** [review]
Conflicts

**Review by Thoralf-M — APPROVED** [review]

File: sdk/src/wallet/error.rs [hunk 7]
```diff
@@ -129,6 +129,12 @@ pub enum Error {
     ImplicitAccountNotFound,
 }
 
+impl Error {
+    pub fn other<E: 'static + std::error::Error + Send + Sync>(err: E) -> Self {
```
Comment (thibault-martinez) [inline sdk/src/wallet/error.rs:133]: Do you want to keep this even if not used atm?

Comment (DaughterOfMars) [reply sdk/src/wallet/error.rs:133]: Hmm. I think so. We have the variant even though we don't use it, but if we do I'm going to want this method to exist. But technically we could add this method at any time without a breaking change so 🤷‍♀️

**Review by thibault-martinez — APPROVED** [review]
